components/push/src/subscriber.rs

```diff
@@ -175,6 +188,108 @@ impl PushManager {
     }
 }
 
+// DB persisted rate limiter.
+// Implementation notes: This saves the timestamp of our latest call and the number of times we have
+// called `Self::check` within the `Self::periodic_interval` interval of time.
+pub struct PersistedRateLimiter {
+    op_name: String,
+    periodic_interval: u64, // In seconds.
+    max_requests_in_interval: u16,
+}
+
+impl PersistedRateLimiter {
+    pub fn new(op_name: &str, periodic_interval: u64, max_requests_in_interval: u16) -> Self {
+        Self {
+            op_name: op_name.to_owned(),
+            periodic_interval,
+            max_requests_in_interval,
+        }
+    }
+
+    pub fn check(&mut self, store: &Store) -> bool {
+        let (timestamp, count) = self.get_counters(store);
+
+        let now = Self::now_secs();
+        if (now - timestamp) >= self.periodic_interval {
+            log::info!(
+                "Resetting. now({}) - {} < {} for {}.",
+                now,
+                timestamp,
+                self.periodic_interval,
+                &self.op_name
+            );
+            let timestamp = Self::now_secs();
+            self.persist_counters(store, timestamp, 0);
+        } else {
+            log::info!(
+                "No need to reset inner timestamp and count for {}.",
+                &self.op_name
+            )
+        }
+
+        // within interval counter
+        if count > self.max_requests_in_interval {
+            log::info!(
+                "Not allowed: innerCount($innerCount) > $MAX_REQUEST_IN_INTERVAL for {}.",
+                &self.op_name
+            );
+            return false;
+        }
+
+        log::info!("Allowed to pass through for {}!", &self.op_name);
+        // Increment the counter and persist it.
+        self.persist_counters(store, timestamp, count + 1);
+
+        true
+    }
+
+    fn db_meta_keys(&self) -> (String, String) {
+        (
+            format!("ratelimit_{}_timestamp", &self.op_name),
+            format!("ratelimit_{}_count", &self.op_name),
+        )
+    }
+
+    fn get_counters(&self, store: &Store) -> (u64, u16) {
+        (
+            store
+                .get_meta(&format!("ratelimit_{}_timestamp", self.op_name))
+                .ok()
+                .flatten()
+                .map(|s| s.parse())
+                .transpose()
+                .ok()
+                .flatten()
+                .unwrap_or_default(),
+            store
+                .get_meta(&format!("ratelimit_{}_count", self.op_name))
+                .ok()
+                .flatten()
+                .map(|s| s.parse())
+                .transpose()
+                .ok()
+                .flatten()
+                .unwrap_or_default(),
```

Comment:
FWIW, I had trouble understanding the sequence of iterator/result operations here. I *think* I have my head around it I wonder if it could be clearer.

One possibility could be to consider a generic version of `get_meta` that has the parsing logic included, something like (untested!):

```
fn get_meta_or_default<T: FromStr + Default>(&self, key: &str) -> T {
  // generic flatten-transpose-etc here.
}
```

Reply:
That was my idea too, but you cannot do generics over numeric types in Rust. I can always cut the `u64` to `u16` though, I'll do that :)